metadata-ingestion/src/datahub/utilities/memory_leak_detector.py

```diff
@@ -0,0 +1,123 @@
+import fnmatch
+import gc
+import logging
+import sys
+import tracemalloc
+from collections import defaultdict
+from functools import wraps
+from typing import Any, Callable, Dict, List, Optional, TypeVar, Union, cast
+
+logger = logging.getLogger(__name__)
+T = TypeVar("T")
+
+
+def _trace_has_file(trace: tracemalloc.Traceback, file_pattern: str) -> bool:
+    for frame_index in range(0, len(trace)):
+        cur_frame = trace[frame_index]
+        if fnmatch.fnmatch(cur_frame.filename, file_pattern):
+            return True
+    return False
+
+
+def _init_leak_detection() -> None:
+    # Initialize trace malloc to track up to 25 stack frames.
+    tracemalloc.start(25)
+    if sys.version_info[0] >= 3 and sys.version_info[1] >= 9:
+        # Nice to reset peak to 0. Available for versions >= 3.9.
```

Comment:
just curious - what exactly does this do?

Reply:
Resets the peak usage to 0. The subsequent call to tracemalloc.get_traced_memory() would report peak usage from this point onwards. This is a good metric to look at what the peak usage has been like.